### ISSUE TITLE
Building OpenNSL

### ISSUE THREAD

**Issue by ggiamarchi** [Broadcom-Switch/OpenNSL#29]
I wish to build OpenNSL from sources. In this repository I see binaries, build instructions but can't figure out where to find full sources.

Do I miss something ?

**Comment by simonknee** [comment]
You are right, this GitHub release only contains binaries.  You can build your OpenNSL applications against the binaries we provide, for the platforms we support or close variants.  If you want to support OpenNSL on another platform that cannot be derived from these binaries then your need your platform vendor to support you, or source code access to OpenNSL.

Source code access requires a Software License Agreement (SLA) with Broadcom.  Reach out to me at simon.knee at broadcom.com if you are interested.

**Comment by davidgengenbach** [comment]
@ggiamarchi If you just want to get OpenNSL working a good starting point is:

- Go into ONIE rescue mode (per serial console), execute:
	- `onie-nos-install http://opennetlinux.org/binaries/2018.06.19.2333.16f97c4cae7fbe1c1d0761ad077b66c3f87bbf0e/ONL-master-ONL-OS-DEB8-2018-06-19.2333-16f97c4-AMD64-INSTALLED-INSTALLER`
	- This will install ONL
- (This will automatically reboot the switch)
- On the reboot, wait for the ONIE "Press CTRL + C to access interactive console" part and press CTRL + C
	- Execute `echo >> /mnt/onl/boot/boot-config; echo "NETAUTO=dhcp" >> /mnt/onl/boot/boot-config`
		- This is to enable DHCP on the managment port
			- (There might be a better way to do this)
				- ONL itself does not execute `dhclient ma1` by itself => no DNS per default 
	- Execute `reboot`
- After reboot, there should be a login prompt for ONL
	- Credentials: `root/onl`
- Execute `wget --no-check-certificate https://support.edge-core.com/hc/en-us/article_attachments/360016901413/opennsl-accton_3.5.0.3_accton5.0-0_amd64.deb`
- Execute `dpkg -i opennsl-accton_3.5.0.3_accton5.0-0_amd64.deb`
- (optional) Execute `ldconfig`
- Execute `opennsl_setup`
	- There might be some output with `permission denied`
	- ... but if there are no `insmod` kernel module errors, you should be fine.

To test the installation, execute `/usr/bin/opennsl-accton/examples/example_stat`. Sample output:
```
root@localhost:/usr/bin/opennsl-accton/examples# /usr/bin/opennsl-accton/examples/example_stat 
Initializing the system.
Platform default configuration is used
Platform Boot flags: 0x0
DMA pool size: REDACTED
PCI unit 0: Dev REDACTED, Rev 0x03, Chip REDACTED, Driver REDACTED

Initializing platform

Found board: as5712-54x
SOC unit 0 attached to PCI device REDACTED
rc: unit 0 device REDACTED
rc: MMU initialized

.........
```

![Screenshot_2019-12-12_02:24:32](https://user-images.githubusercontent.com/2621491/70674531-8f15fe00-1c86-11ea-8e4c-862589f4d355.png)

Note: I am very new to SDN or networking in general, so this might not be the best solution. I tried numerous approaches: different kernel module binaries, different kernels, trying to build OpenNSL from source, ... This was the only approach which worked for me.

Offtopic: personally, I think that the state of the (SDN) networking ecosystem is *very, very* far from perfect: the documentation on most software is either outdated or incomplete. There are binaries floating around with any meaningful context. The paths to published images change, which means that links in forum posts lead to 404s.
Even to get the most basic examples running on real popular hardware is a complete pain and requires knowing a myriad of technologies, hidden commands, pitfalls en masse, ...
Also: *Open*NSL might also be interpreted as a misnomer if the source is not open.

> You are right, this GitHub release only contains binaries. You can build your OpenNSL applications against the binaries we provide, *******for the platforms we support or close variants*******. If you want to support OpenNSL on another platform that cannot be derived from these binaries then your need your platform vendor to support you, or source code access to OpenNSL.

I thought the main aspiration/motivation of moving to SDNs was **exactly** to circumvent vendors defining which hardware is supported how and when.
One of the biggest changes of SDN with the open-source approach is that compatibility can also be ensured by third-parties.
